index.bs

```diff
@@ -1880,23 +1888,54 @@ BrowsingContextEvent = (
 browsingContext.BrowsingContext = text;
 </pre>
 
-Each [=/browsing context=] has an associated <dfn export>browsing context
-id</dfn>, which is a string uniquely identifying that browsing context. This is
-implicitly set when the context is created. For browsing contexts with an
-associated WebDriver [=window handle=] the [=/browsing context id=] must be the
-same as the [=window handle=].
+A [=/navigable=] is associated with a <dfn export>context id</dfn>, which is a
+string uniquely identifying that browsing context within a specific [=bidi
+session=].
+
+A [=BiDi session=] has a <dfn>context id map</dfn> which is a weak map between
+[=navigables=] and their corresponding [=context id=]. When the session is
+created this map is empty.
 
 <div algorithm>
-To <dfn>get a browsing context</dfn> given |context id|:
+To <dfn>get context id</dfn> given |session| and |context|:
+
+1. Let |navigable| be |context|'s [=active document=]'s [=node document=]'s
+   [=node navigable=].
+
+1. If |navigable| is null, return [=error=] with [=error code=] [=no such
+   frame=].
+
+   TODO: Once we change the callers to use navigables throughout the condition
+   will become unreachable.
+
+1. Let |context id map| be |session|'s [=context id map=].
+
+1. If |context id map| doesn't [=map/contain=] |navigable|:
+
+  1. If |navigable| is a [=navigable/top-level traversable=] and |session| is a
+     [=HTTP session=], let |context id| be the [=window handle=] for
+     |navigable|.  Othereise, let |context id| be a new, unique, string.
+
+  1. [=map/Set=] |context id map|[|navigable|] to |context id|.
+
+1. Return [=success=] with data |context id map|[|navigable|].
+
+</div>
+
+<div algorithm>
+To <dfn>get a browsing context</dfn> given |session| and |context id|:
 
 1. If |context id| is null, return [=success=] with data null.
 
-1. If there is no browsing context with [=browsing context id=] |context id| return
-   [=error=] with [=error code=] [=no such frame=]
+1. Let |context id map| be |session|'s [=context id map=].
+
+1. For each |navigable| → |id| of |context id map|:
```

Comment:
Note that WeakMaps do not support enumeration and as such there is currently no way to actually retrieve the navigable corresponding to the `context id` at the moment.